rfcs/45-stageless.md

```diff
@@ -0,0 +1,759 @@
+Of particular note:
+
+- Stages entangle ordering and command processing.
+- Plugin authors have no tools to export their systems other than wrapping them in stages, but the scheduling of imported stages can't be controlled by the user.
+- Systems and system sets cannot have multiple run criteria.
+- There's no clear pattern for users to architect turn-based games.
+- There's no clear pattern for users to make a state "span across" multiple stages with our state machine model.
+- Our state machine model (a stack) doesn't enable enough use cases to warrant its complexity.
+- There are just too many types and methods. (looking at you, "`.add_system_set_to_startup_stage`")
+
+Unfortunately, all of these problems are deeply interwoven. Despite our best efforts, fixing them incrementally at either the design or implementation stage is impossible, as it results in myopic architecture choices and terrible technical debt.
+
+## User-facing explanation
+
+### Scheduling terminology and overview
+
+Let's define some terms so that hopefully we're all on the same page.
+
+- **system**: a stateful instance of a function that accesses data stored in a world
+- **system set**: a system label that represents a group of systems and other sub-sets
+- **condition**: a special kind of system that guards the execution of a system (set)
+- **dependency**: a system (set) that must complete before the system (set) of interest can run
+- **scheduler**: the programmer, who is attempting to control when and under what conditions systems run
+- **schedule** (verb): specify when and under what conditions systems run
+- **schedule** (noun): the executable form of a system set
+- **executor**: executes a schedule on a world
+- **"ready"**: when a system is no longer waiting for dependencies to complete
+
+To build a Bevy app, users have to specify when their systems should run. By default, systems have no strict order nor conditions for execution. The process of specifying those is called **scheduling**. To make things more ergonomic, systems can be grouped into **system sets**, which can be ordered and conditioned in the same manner as systems. Furthermore, systems and system sets can be ordered together and grouped together *within larger sets*, which allows users to describe logical hierarchies.
+
+In short, for any system or system set, users can:
+
+- add execution order constraints relative to other systems or sets (e.g. "this system runs before A")
+- add conditions that must be true for it to run (e.g. "this system only runs if a player has full health")
+- add it to sets (which define common behavior for all systems within that set)
+
+These properties are all additive.
+Adding another does not replace an existing one, and they cannot be removed.
+
+Example:
+
+```rust
+use bevy::prelude::*;
+
+#[derive(State)]
+enum GameState {
+    Playing,
+    Paused
+}

```

Comment:
The current bevy state system uses the discriminant to differentiate between states which means you can't have a state with a payload because it is never considered:

```rs
struct PlayerId {
    /* fields */
}

enum GameState {
    Turn(PlayerId),
    Paused,
}
```

Will this implementation allow for more complicated state types?